assembly/src/errors.rs

```diff
@@ -81,13 +81,16 @@ pub enum AssemblyError {
     },
     #[error(transparent)]
     #[diagnostic(transparent)]
-    Other(#[from] RelatedError),
-    #[error(transparent)]
-    Forest(#[from] MastForestError),
+    Other(RelatedError),
+    // Technically MastForestError is the source error here, but since AssemblyError is converted
+    // into a Report and that doesn't implement core::error::Error, treating MastForestError as a
+    // source error would effectively swallow it, so we include it in the error message instead.
+    #[error("{0}: {1}")]
+    Forest(&'static str, MastForestError),
```

Comment:
Hmmm, so is miette's `Report` incompatible with `Error::source`? Or is this a by-product of the `miden-miette` situation described in the issue somehow?

Also `MastForestError` currently doesn't use any `#[source]` - but if we add one, then it won't get displayed by `Report`, right? Isn't this a larger issue?

Reply:
Yeah, `Report` doesn't implement `core::error::Error` (https://github.com/zkat/miette/issues/366), so anything that is not part of the `Display` impl of an error would be silently swallowed.

This makes me think that we might not want to return `Report` from library functions like `assemble_*`, but the underlying `AssemblyError` instead. On the other hand, this will usually mean that no Report is constructed for such errors and so the error is worse than it could be.
This is why for now, I'm including the printed report in the display impl of an error that wraps a `Report`, e.g.: https://github.com/0xPolygonMiden/miden-base/blob/67272131f02ed73b5b919223610e97afa7bdc8db/objects/src/errors.rs#L29
(The other reason being that `Report` doesn't implement `core::error::Error`).

I'm not sure if I'm missing something here, since the interop between `core::error::Error` and `Report` feels unsatisfying.

Reply:
@bitwalker IIRC you'd like us to return `Report` much more, right? From previous conversations, I think you even suggested we convert `AssemblyError` to just use `Report`s everywhere?

I haven't played with `miette` enough to have a good intuition here,  but typically when all your error are user-facing/to be printed only, you'd want to use `anyhow`/`eyre`/`miette` (instead of an `enum` that you end up printing out anyways). We could do that, but then I'm still not sure how we would "bridge properly" between e.g. errors coming from `miden-processor` which would all use `Error::source`.

TLDR this PR is already a big upgrade from the current state of things, but I wonder what "optimal strategy" we should be striving for here.

Reply:
Yeah - we want to remove `AssemblyError`. There is an https://github.com/0xPolygonMiden/miden-vm/issues/1431 specifically for this. I think we can address this right after this PR.

Reply:
The idea of `Report` (or alternatively, something like `anyhow::Error` if there isn't any associated source code, which is where `Report` really shines) is to allow ad-hoc diagnostics to be constructed/created, while handling all of the annoying/boilerplatey work of conversion/rendering/etc. The use of ad-hoc diagnostics ends up producing much more useful errors in practice IMO.

The downside of using `Report` (or `anyhow::Error`) is that you can't easily obtain the concrete underlying error type (if there is one), in order to say, assert that it is of a specific type. I don't find this particularly important IMO, and there are better ways to test errors (e.g. how we test assembler errors), and downcasting can be used in many situations if really needed.

You can wrap `Report` in a type that allows `#[from]` or `#[source]` conversion, IIRC that is what the `RelatedError` type exists for, but it is probably better to _not_ have a variant that is just a transparent wrapper around `Report`.